### PR TITLE
Bluetooth: host: Fix missing net_buf_unref

### PR DESCRIPTION
**Pull request by LingaoM**
When call `bt_conn_send_cb` with conn state not connect,
will cause buffer unref.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

### CODE REVIEW

**Comment by Thalley** [comment]
From the documentation of the function: 

```
/* Send data over a connection
 *
 * Buffer ownership is transferred to stack in case of success.
 *
 * Calling this from RX thread is assumed to never fail so the return can be
 * ignored.
 */
int bt_conn_send_cb(struct bt_conn *conn, struct net_buf *buf,
		    bt_conn_tx_cb_t cb, void *user_data);
```

**Buffer ownership is transferred to stack in case of success.** => I think this is intentional, and that the stack should not unref teh buffer in case of error.

**Comment by jhedberg** [comment]
@Thalley correct, or then we change the semantics of the function and make sure that all callers update their expectations regarding buffer ownership (not sure if we want that - there's probably a reason why caller ownership is retained for error situations).

Btw, considering the quoted "Calling this from RX thread is assumed to never fail so the return can be ignored" part of the documentation. Is this still true and accurate after #40273 was merged? @LingaoM any comment?

File: subsys/bluetooth/host/conn.c [hunk 1]
```diff
@@ -372,28 +372,32 @@ static struct bt_conn_tx *conn_tx_alloc(void)
...
 	if (conn->state != BT_CONN_CONNECTED) {
 		BT_ERR("not connected!");
-		return -ENOTCONN;
+		err = -ENOTCONN;
+		goto end;
 	}
 
 	if (cb) {
 		tx = conn_tx_alloc();
 		if (!tx) {
```
Comment (LingaoM) [inline subsys/bluetooth/host/conn.c:389]: @Thalley @jhedberg I think it's still possible to return failure here. For example, call write command or notify in syswork. What happens when free TX space is insufficient?

``` C
static struct bt_conn_tx conn_tx[CONFIG_BT_CONN_TX_MAX];
```

Comment (Thalley) [reply subsys/bluetooth/host/conn.c:389]: How do you mean?

We currently already return `ENOBUFS` if there are no TX contexts left.

Comment (LingaoM) [reply subsys/bluetooth/host/conn.c:389]: But in this case, the buffer is still occupied without unref.

Comment (LingaoM) [reply subsys/bluetooth/host/conn.c:389]: We lack the consideration of unref under abnormal circumstances, which is the purpose of my pr.

**Comment by alwa-nordic** [comment]
Closing: Not a bug since implementation matches documentation. See https://github.com/zephyrproject-rtos/zephyr/pull/44922#issuecomment-1102296411.

Consider opening a ticket for an API change if you want to get this change in.

**Comment by Thalley** [comment]
> Closing: Not a bug since implementation matches documentation. See [#44922 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/44922#issuecomment-1102296411).
> 
> Consider opening a ticket for an API change if you want to get this change in.

Just to be clear: This is an internal function and not part of a public API.

**Pull request closed without merging**